### PR TITLE
Preserve format when copy/paste yaml file

### PR DESCRIPTION
**Pull request by TheGostKasper**
Closes [Preserve my formatting when I copy/paste excerpts from WGE UI yaml files](https://github.com/weaveworks/weave-gitops-enterprise/issues/3321)

**What changed?**
Add a copy button to `YamlView` component

I updated `CopyToClipboard` component to accept different option ( icon , icon + text ) and added it to `YamlView` component

https://github.com/weaveworks/weave-gitops/assets/4614360/860dde7d-9825-4dde-8bde-95dea8beeb0c

https://github.com/weaveworks/weave-gitops/assets/4614360/06f3a091-214a-4e81-add0-f3aa3f245659

### CODE REVIEW

**Comment by TheGostKasper** [comment]
@mmoulian @LappleApple 
I went for showing a copy Icon  instead of `icon+text`, if you have any comments or you want the other option, please let me know .

**Comment by LappleApple** [comment]
Works for me @TheGostKasper -- thanks for jumping on this.

**Comment by opudrovs** [comment]
I've tested this branch and verified that copying YAML with the Copy button copies valid YAML and preserves the indents and does not insert extra space between the lines. ✨ 

But the acceptance criteria mention copying excerpts:

> As a platform dev, I would like to be able to preserve my line spacing and formatting when I copy-paste excerpts from the yaml file in the yaml tab, so that I don't insert errors by mistake or have to redo work.

And this comment from @jpellizzari mentions copying highlighted text but also mention adding a Copy button as an alternative (what this PR does):
https://github.com/weaveworks/weave-gitops-enterprise/issues/3321#issuecomment-1715929216

When I highlight portions of YAML on the web page and copy it with Ctrl+C, this is what I see on paste:

Screenshot 2023-09-13 at 18 09 28

Indentation and space between lines are not preserved.

So, @LappleApple @jpellizzari could you please clarify if the acceptance criteria should be edited to only require copying the whole YAML with the Copy button or if the current PR should also include the "copy highlighted text and preserve indentation" functionality?

If only the former, I can approve this PR. If the latter is required too, more work is needed. In principle, this issue could be split into two issues, because the first part is much easier to implement than copying highlighted text with preserving formatting.

**Comment by opudrovs** [comment]
@jpellizzari OK, fine with me!

**Comment by LappleApple** [comment]
@jpellizzari @opudrovs Hi both, I didn't have time to discuss implementation details or add Acceptance Criteria before this got picked up via UI triage and over to Timberwolf. Great to see it moving forward. If you need me to create ACs let me know, but it looks like Jordan's got this?

**Comment by TheGostKasper** [comment]
@opudrovs , the issue here is with `showLineNumbers` as it displays the content `code` in a flex way with custom style to the number section + code which require to add onCopy event globally to maintain the content to re-format it to the correct format (regex will be used , or customize the line style) --> **Do we really need to do it!?**

image

On the other hand removing `showLineNumbers` fron `SyntaxHighlighter` will display the content as
image



What options we have here ?
- Remove `showLineNumbers` to preserve text format on Copy 
- Show a Copy icon to copy content 
- Customize style or use Reqex to preserve format for the copied lines 
- others... ?


The easiest way for now was to show the Icon .

**Comment by opudrovs** [comment]
@TheGostKasper hmm, interesting. Would it be possible to, maybe, display line numbers in some common margin (maybe with some before pseudo-elements) but position them absolutely relative to each line span and make them non-selectable? 

Or if it's overcomplicated (probably it is), the decision depends on the Product's priorities: what is more important, displaying line numbers or being able to select portions of the YAML. Either option is fine with me.

The third option (using Regexes to clean up copied text) sounds interesting too and might the least troublesome if the copied text is consistent enough.

In principle, I would prefer to switch to another YAML/markdown/rich text component (maybe a third-party solution) which supports selecting and copying code without line numbers. It should be such a standard task. But I am not sure how much time searching for the right component will take, like if it's worth it.

**Comment by TheGostKasper** [comment]
@opudrovs changing lineProps to use `textWrap` instead of `flexWrap` solves it

**Comment by opudrovs** [comment]
@TheGostKasper great, I'll re-test it now.

Can you please update the failing snapshot test?

File: ui/components/YamlView.tsx [hunk 1]
```diff
@@ -53,33 +53,37 @@ function UnstyledYamlView({
       },
     },
 
-    lineProps: { style: { flexWrap: "wrap" } },
+    lineProps: { style: { TextWrap: "wrap" } },
```
Comment (opudrovs) [inline ui/components/YamlView.tsx:56]: ❓ Why is `Text` capitalized in `TextWrap`?

**Comment by opudrovs** [comment]
Re-tested it manually, works fine for me in Chrome. Great job! ✨ 

The only minor issue left is that when selecting and copying a portion of YAML in Firefox, blank lines are inserted between YAML lines.

This is what I see when copy=pasting several lines from Chrome (works as expected):

```
metadata:
  annotations:
    metadata.weave.works/CreatedBy: Value 4
    metadata.weave.works/Version: some version
    metadata.weave.works/created-by: Value 2
    metadata.weave.works/createdBy: Value 3
    metadata.weave.works/description: Value 1
    metadata.weave.works/html: <p><b>html</b></p>
    metadata.weave.works/link-to-google: https://google.com
    metadata.weave.works/multi-lines: |
      This is first line
      This is second line
  creationTimestamp: 2023-09-15T15:03:27Z
  finalizers:
    - finalizers.fluxcd.io
  generation: 1
  labels:
    kustomize.toolkit.fluxcd.io/name: flux-system
    kustomize.toolkit.fluxcd.io/namespace: flux-system
```

and this is what I see when copy-pasting from Firefox:

```
metadata:

  annotations:

    metadata.weave.works/CreatedBy: Value 4

    metadata.weave.works/Version: some version

    metadata.weave.works/created-by: Value 2

    metadata.weave.works/createdBy: Value 3

    metadata.weave.works/description: Value 1

    metadata.weave.works/html: <p><b>html</b></p>

    metadata.weave.works/link-to-google: https://google.com

    metadata.weave.works/multi-lines: |

      This is first line

      This is second line

  creationTimestamp: 2023-09-15T15:03:27Z

  finalizers:

    - finalizers.fluxcd.io

  generation: 1

  labels:

    kustomize.toolkit.fluxcd.io/name: flux-system

    kustomize.toolkit.fluxcd.io/namespace: flux-system
```

But I think it can be considered a Firefox issue.

**Review by opudrovs — APPROVED** [review]
LGTM! 🎉

**Pull request merged**